### ISSUE TITLE
Dynamic scrolling through a custom Element

### ISSUE THREAD

**Issue by nots1dd** [ArthurSonzogni/FTXUI#984]
Hi, I have been using FTXUI for some time now and am trying to render my own scroll like so:

```cpp
auto RenderSongMenu(const std::vector<Element>& items, int* selected_index, TrueColors::Color color, size_t max_visible_items = 30)
{
    Elements rendered_items;

    if (items.empty()) {
        return vbox(text("No songs available"));
    }

    *selected_index = std::max(0, std::min(*selected_index, (int)items.size() - 1));

    size_t start_index = 0;
    size_t end_index = std::min(max_visible_items, items.size()); // Default to the first `max_visible_items`.

    if (*selected_index >= (int)(start_index + max_visible_items)) {
        start_index = *selected_index - max_visible_items + 1;
        end_index = std::min(start_index + max_visible_items, items.size());
    } else if (*selected_index < (int)start_index) {
        start_index = *selected_index;
        end_index = std::min(start_index + max_visible_items, items.size());
    }

    for (size_t i = start_index; i < end_index; ++i) {
        bool is_selected = (i == *selected_index);
        auto style = is_selected ? getTrueColor(color) : nothing;
        auto inverted_style = is_selected ? inverted : nothing;
        rendered_items.push_back(items[i] | style | inverted_style);
    }

    auto menu = vbox(std::move(rendered_items));
    return menu;
}
```

this function is called like so:

```cpp
  void CreateComponents()
  {
    MenuOption artist_menu_options;
    artist_menu_options.on_change = [&]()
    {
      if (focus_on_artists && selected_artist < current_artist_names.size())
      {
        UpdateSongsForArtist(current_artist_names[selected_artist]);
      }
    };
    artist_menu_options.focused_entry = &selected_artist;

    MenuOption song_menu_options;
    song_menu_options.on_change     = [&]() {};
    song_menu_options.focused_entry = &selected_inode;

    artists_list = Menu(&current_artist_names, &selected_artist, artist_menu_options);
    songs_list   = Renderer(
      [&]() mutable
      {
        return RenderSongMenu(current_song_elements, &selected_inode,
                                TrueColors::Color::LightYellow); // This should return an Element
      });

    auto main_container = Container::Horizontal({artists_list, songs_list});
... // continues
```

this works but I want the scroll to be dynamic.

I have taken a look at the Scroller component that was present in git-tui using FTXUI however I found it pretty hard to integrate with my code so I tried to create a custom Element that obeys the parent container's CatchEvent and Renderer

Any tips or references regarding how to implement fluid scrolling to a custom Element (similar to Scroller) or atleast getting the current terminal's dimensions through FTXUI would be great

**Comment by ArthurSonzogni** [comment]
>  or atleast getting the current terminal's dimensions through FTXUI would be great

```cpp
ftxui::Terminal::Size().dimx
ftxui::Terminal::Size().dimy
```
Does that help?

About `Renderer`, I would suggest passing the `artists_list` as the first argument of `Renderer` to "connect" it to the main component. If you don't, events won't be properly processed by the `Menu`. You would renderer a "dead" component.

**Comment by nots1dd** [comment]
> Hello @nots1dd
> 
> > or atleast getting the current terminal's dimensions through FTXUI would be great
> 
> ```c++
> ftxui::Terminal::Size().dimx;
> ftxui::Terminal::Size().dimy;
> ```
> 
> Does that help?
> 
> About `Renderer`, I would suggest passing the `artists_list` as the first argument of `Renderer` to "connect" it to the main component. If you don't, events won't be properly processed by the `Menu`. You would renderer a "dead" component.

Can you elaborate more on this? I'll send more code snippets on this:

```cpp
    artists_list = Menu(&current_artist_names, &selected_artist, artist_menu_options);
    songs_list   = Renderer(
      [&]() mutable
      {
        return RenderSongMenu(current_song_elements, &selected_inode,
                                TrueColors::Color::LightYellow); // This should return an Element
      });

    auto main_container = Container::Horizontal({artists_list, songs_list});

    /* Adding DEBOUNCE TIME (Invoking PlayCurrentSong() too fast causes resources to not be freed)
     */

    auto last_event_time = std::chrono::steady_clock::now(); // Tracks the last event time
    int  debounce_time = std::stoi(std::string(parseTOMLField(PARENT_DBG, "debounce_time_in_ms")));
    if (debounce_time < 500)
      debounce_time = 500; // min debounce time is 0.5s
    const int final_debounce_time = debounce_time;
    auto      debounce_duration   = std::chrono::milliseconds(final_debounce_time);

    main_container |= CatchEvent(
      [&](Event event)
      {
        auto is_keybind_match = [&](char key) -> bool
        {
          return (event.is_character() && event.character() == std::string(1, key)) ||
                 (event == Event::Special(std::string(1, static_cast<char>(key))));
        };

        if (active_screen == SHOW_HELP_SCREEN)
        {
          if (event.is_character() && (event.character()[0] == global_keybinds.show_help ||
                                       std::toupper(global_keybinds.quit_app) ||
                                       event.character()[0] == global_keybinds.quit_app))
```

The function goes on to go to every keybind i have set up till I set my renderer like:

```cpp
/* In my UI class... */
private:
Component renderer;

/* Other functions */

// Inside CreateComponents()...

    renderer =
      Renderer(main_container,
               [&]
               {
                 int   duration = GetCurrentSongDuration();
                 float progress = duration > 0 ? (float)current_position / duration : 0;

                 Element interface;
                 if (active_screen == SHOW_HELP_SCREEN)
                 {
                   interface = RenderHelpScreen();
                 }
                 else
                 {
                   interface = RenderMainInterface(progress);
                 }

                 if (show_dialog)
                 {
                   // Create a semi-transparent overlay with the dialog box
                   interface = dbox({
                     interface,              // Dim the background
                     RenderDialog() | center // Center the dialog both horizontally and vertically
                   });
                 }
                 if (active_screen == SHOW_LYRICS_SCREEN)
                 {
                   interface = RenderLyricsAndInfoView();
                 }

                 return vbox(interface);
               });
```

This renderer is then called in the main function of the class like:

```cpp
  void Run()
  {
    auto        screen = ScreenInteractive::Fullscreen();
    std::thread refresh_thread(
      [&]
      {
        while (!should_quit)
        {
          using namespace std::chrono_literals;
          UpdateVolume();
          std::this_thread::sleep_for(0.1s);

          if (is_playing)
          {
            current_position += 0.1;
            if (current_position >= GetCurrentSongDuration())
            {
              PlayNextSong();
            }
          }
          screen.PostEvent(Event::Custom);
        }
      });

    refresh_thread.detach();
    screen_ = &screen;
    screen.Loop(renderer);
  }
```

Please give your input on this type of implementation, I have read through FTXUI source code, documentation and projects that use FTXUI and honestly speaking I am not the best designer so I struggle on creating designs in an optimal approach

Regarding the `ftxui::Terminal::Size()` functions I will try and take a look into them.

**Comment by nots1dd** [comment]
> Hello @nots1dd
> 
> > or atleast getting the current terminal's dimensions through FTXUI would be great
> 
> ```c++
> ftxui::Terminal::Size().dimx;
> ftxui::Terminal::Size().dimy;
> ```
> 
> Does that help?
> 
> About `Renderer`, I would suggest passing the `artists_list` as the first argument of `Renderer` to "connect" it to the main component. If you don't, events won't be properly processed by the `Menu`. You would renderer a "dead" component.

Oh I see, this is using ioctl under the hood, I tried doing it this way but I assumed I could get the height of every component / element in FTXUI with respect to the terminal

For example in FTXUI we can set the `HEIGHT` enum in `size()` like so:

```cpp
size(HEIGHT, EQUAL 10)
``` 

This would set the number of lines (or elements) in that Element as 10 (that constitutes as height)

So I assumed we could also find the height of every Element as we can set them in this manner

At the end of the day, I only wanted this metric to allow for my implementation for dynamic scrolling, so if you have any workarounds or a better way of making `RenderSongMenu()` a scrollable element, I would really appreciate it :)

**Comment by nots1dd** [comment]
Ok i fixed this issue it was pretty simple actually I was overdoing it

I modified the Scroller component a little (specifically the OnEvent function) and used it like so:


```cpp
/* Calling the Scroller in my UI handler class */
    INL_Component_State.songs_list   = Scroller(Renderer(
      [&]() mutable
      {
        return RenderSongMenu(current_song_elements); // This should return an Element
      }), &selected_inode, global_colors.menu_cursor_bg);

// INL_Component_State is a struct with all the Components required for the UI lifetime
```

```cpp
/* In another header RenderSongMenu's definition and logic */
auto RenderSongMenu(const std::vector<Element>& items)
{
  Elements rendered_items;
  for (int i = 0; i < items.size(); ++i)
  {
    rendered_items.push_back(items[i] | frame);
  }

  return vbox(std::move(rendered_items));
}
```

Now I simply call `INL_Component_State.songs_list` like so:

```cpp
    auto right_pane =
      vbox({
        text(" Songs") | bold | color(global_colors.songs_title_bg) | inverted,
        separator(),
        INL_Component_State.songs_list->Render() | frame | flex | getTrueColor(TrueColors::Color::White),
      }) |
      borderHeavy | color(GetCurrWinColor(!focus_on_artists));

    auto panes = vbox({hbox({
                         left_pane | size(WIDTH, EQUAL, 100) | size(HEIGHT, EQUAL, 100) | flex,
                         right_pane | size(WIDTH, EQUAL, 100) | size(HEIGHT, EQUAL, 100) | flex,
                       }) |
                       flex}) |
                 flex;
//... so on this logic is returned as an Element
// This element was then rendered finally in screen.Loop() callback with custom keybinds and // CatchEvent that **SHOULD NOT** depend on Scroller Component's custom OnEvent (it will
// conflict and give unexpected results.
```

I believe adding `frame` is essential to all components to render properly (like Menu, etc.) else this implementation will not work.
